### ISSUE TITLE
Remotecall and @spawnat for threads?

### ISSUE THREAD

**Issue by ChrisRackauckas** [JuliaLang/julia#17933]
[Is there a good answer to this SO question?](http://stackoverflow.com/questions/36262554/how-do-i-call-function-in-a-separate-thread-in-julia-0-5-dev). Distributed parallelism (multiprocessing) has `@spawnat` and remotecall. Could similar functions be available for spawning a function call on a thread?

**Comment by oxinabox** [comment]
Transcribing this from other places. (SO, Gitter)

I spent a while digging around in the internals, so see if I could make this work as a hack.

 - Use async to run a function on all threads using `jl_threading_run`, 
 - Yield then terminate (return) if the current thread is thread 1. (to free the async call from waiting for the ccall of `jl_threading_run` from finishing)
 - If you are otherwise not the tartget thread, just terminate (return)
 - if you are the target thread, run that actual function, storing the result in a `Future`

See https://gist.github.com/oxinabox/d0eb8a126948a77680fd6ba08582eca8


But I couldn't quiet get it to work. Not entirely sure why.
It is not a great solution anyway, since it fires work on to threads it doesn't want to use anyway.


I think to do what I wanted to do you actually need to always be inside a eternally looping jl_threading_run (even for your "main" thread), and have each thread polling/waiting for work -- like setting up a work stealing architecture.

**Comment by kpamnany** [comment]
The current parallel loop infrastructure is not suitable for this; we plan to introduce an ``@spawn`` macro in the future. Much work needs to happen in simplifying and unifying coroutines, threads, locks, workers...

In the meantime, http://docs.julialang.org/en/latest/manual/parallel-computing/#threadcall-experimental may suit your needs.

**Comment by tknopp** [comment]
This would be great to have since it makes asynchronous programming in GUI applications (e.g. Gtk.jl) easier. Here I kind of like the `BackgroundWorker` model from C# where after the thread has finished a callback in the main thread is executed.

**Comment by ancapdev** [comment]
Adding another use case to express interest in this.

I'm embedding Julia in a soft-realtime event processing system, where from time to time the system is parameterized in a new way and we take advantage of Julia's code generation facilities to generate event handlers specific to the new parameterization. These event handlers are later invoked for the time critical processing. The parameterization doesn't change terribly often, but the code generation and JIT is something I'd much rather be able to push to a background thread, so the main thread can continue servicing events using the old parameters.

Even some guidelines for how to hack something together with the current threading functionality would be very helpful. With the parallel fork/join pattern that's currently available, it looks like some of the fundamentals (like concurrent GC and JIT) are all available, or maybe I'm missing something?

**Comment by yuyichao** [comment]
> it looks like some of the fundamentals (like concurrent GC and JIT) are all available

None of the two are available.

**Comment by yuyichao** [comment]
No. All threads will wait for the GC to finish if it needs to run.

**Comment by vitreo12** [comment]
Are there any news about this? Is this set to be released on a specific future Julia version (1.5, 2.x, 3.x....)?

**Comment by oxinabox** [comment]
1.x after sometime after the first PATR PR is merged. Or as part of that.
(Or so I believe)
